### DISCUSSION TITLE
How to encode a custom CLR object to a PyObject in a codec?

### DISCUSSION THREAD

**Discussion by ophirKatz** [pythonnet/pythonnet#2080] [Q&A]
### Environment

-   Pythonnet version: 3.0.1
-   Python version: 3.11
-   Operating System: Win11
-   .NET Runtime: .net core (.net 7)

### Details

-   Describe what you were trying to get done.

I'm trying to pass a complex object or enum from python to .net but get an exception:
System.InvalidCastException: 'Unable to cast object of type 'Python.Runtime.CLRObject' to type 'Python.Runtime.ManagedType'.'

-   What commands did you run to trigger this issue? If you can provide a
    [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
    this will help us understand the issue.

```cs

public class SB
{
    public string Text { get; private set; }

    public SB(string text)
    {
        Text = text;
    }

    public void Append(SB other)
    {
        Text += other.Text;
    }
}

public class SBCodec : IPyObjectEncoder, IPyObjectDecoder
{
    public static void Setup()
    {
        PyObjectConversions.RegisterDecoder(new SBCodec());
        PyObjectConversions.RegisterEncoder(new SBCodec());
    }

    public bool CanDecode(PyType objectType, Type targetType)
    {
        return targetType == typeof(SB);
    }

    public bool TryDecode<T>(PyObject pyObj, [UnscopedRef] out T? value)
    {
        var text = pyObj.GetAttr("Text").As<string>();
        value = (T)(object)new SB(text);
        return true;
    }

    public bool CanEncode(Type type)
    {
        return type == typeof(SB);
    }

    public PyObject? TryEncode(object value)
    {
        // what to do here???
        throw new NotImplementedException();
    }
}

Runtime.PythonDLL = "python311.dll";
PythonEngine.Initialize();
SBCodec.Setup();

var filename = "path_to_script\\test.py";

using (Py.GIL())
using (var scope = Py.CreateScope())
{
    var code = File.ReadAllText(filename);
    var compiledScript = PythonEngine.Compile(code);
    scope.Execute(compiledScript);

    var sb = new SB("Hello");
    var sbPython = sb.ToPython();
    
    dynamic func = scope.Get("func");

    func(sbPython);
}

```

```python
from pythonnet import load

load("coreclr")

import clr
clr.AddReference("PythonNet.Test")

from PythonNet.Test import SB

def func(sb):
    sb2 = SB("World")
    sb.Append(sb2)
```

-   If there was a crash, please include the traceback here.

On the line of passing sb2 to the Append method, I'm getting the following exception in my .net call to func

System.InvalidCastException: 'Unable to cast object of type 'Python.Runtime.CLRObject' to type 'Python.Runtime.ManagedType'.'

I have no idea how I should construct a python object in the codec for the conversion, and honestly, this subject's documentation is pretty scarce so I can't find any examples of encoding a clr object.

Please help!

**Comment by ophirKatz** [comment]
By the way, enums don't work out of the box as well when passing to dotnet via a clr method. Do I need the register an encoder for enums as well? There's a built in one but it is marked as obsolete 😕

**Comment by lostmsu** [comment]
When you construct .NET object in Python, I believe you get the object without any codecs applied. In the scenario you provided I don't even see why you'd need a codec in the first place. You can just work with instances of `SB` directly from Python. Codecs are needed when you want to map Python types to .NET types and vice versa (e.g. `ndarray` and `System.Array`). In your case there's no Python type to map to.

**Reply by ophirKatz** [reply]
And still, I'm getting this error while using all .NET objects and types.
The line `sb.Append(sb2)` introduces the error, in which I'm interacting with an existing CLR object and using a method to send another CLR object to it. Yet I'm getting an InvalidCastException, but where am I trying to cast a CLR object to a ManagedType?

**Reply by filmor** [reply]
Have you tried just calling `func` with `sb` instead of `sbPython`?

**Reply by ophirKatz** [reply]
I did... :(

**Reply by filmor** [reply]
The output being? :) I'll try to reproduce this tomorrow, but I don't have a laptop with me right now :)

**Reply by ophirKatz** [reply]
Same... Can't cast a CLR object to a ManageType

**Comment by filmor** [comment]
Well, one issue I can definitely see is that you are doing `pythonnet.load` in an embedded Python.NET (i.e. running from .NET already). This will likely lead to it loading a second instance of the Python.NET assembly and is something that we should prevent. If I remove that and all of the other conversions (so codec and `.ToPython`) the code runs through as expected, it "just" blocks in the end.

**Reply by filmor** [reply]
Ah, yes, that's just because you are missing a `Shutdown()`. Adjusting that everything works as expected (without a codec).

**Reply by ophirKatz** [reply]
That worked! Thank you!
I guess It's just a matter of documentation. It was not clear how to pass CLR object parameters to a python method, and the use case of loading the clr from a python code loaded by Python.NET was not covered as well.
I think I'll add a PR sometimes to address that - it will add some documentation from our experience to the repo. Maybe you'll add that to the website later :)

**Reply by filmor** [reply]
The website is generated from this repo, so all information that you add to `/doc` will appear there. The reason why we don't document much about passing parameters around is that it's really simple in the happy case. Your case in particular is just something that we need to prevent code-wise. The `pythonnet.load` path didn't exist before 3.0, so the code for ".NET in Python" and "Python in .NET" usage could be identical without issues.
